- name: Test include_role
  hosts: testhost

  vars:
    role_name: role1
    test_var: templating test in playbook
    role_vars:
      where_am_i_defined: in the playbook
    entire_task:
      include_role:
        name: role1

  tasks:
    - name: Test basic role include
      include_role:
        name: role1

    - name: Assert that basic include works
      assert:
        that:
          - _role1_result.msg == 'In role1'

    # FIXME This looks like it's running all the roles each time in the loop
    - name: Test role include with a loop
      include_role:
        name: "{{ item }}"
      register: loop_test
      with_items:
        - role1
        - role2
        - role3

    - name: Test including a task file from a role
      include_role:
        name: role1
        tasks_from: tasks.yml

    - name: Test including vars file and tasks file from a role
      include_role:
        name: role3
        tasks_from: vartest.yml
        vars_from: role3vars.yml
        private: no

    - debug:
        msg: "{{item}}"
      with_items:
        - role3_main
        - role3_default
        - role3_var

    - name: Test using a variable for role name
      include_role:
        name: "{{ role_name }}"

    - name: Pass variable to role
      include_role:
        name: role1
        tasks_from: vartest.yml
      vars:
        where_am_i_defined: in the task

    ## FIXME Currently failing
    # - name: Pass all variables in a variable to role
    #   include_role:
    #     name: role1
    #     tasks_from: vartest.yml
    #   vars: "{{ role_vars }}"

    - name: Pass templated variable to a role
      include_role:
        name: role1
        tasks_from: vartest.yml
      vars:
        where_am_i_defined: "{{ test_var }}"

    ## FIXME This fails with the following error:
    ## The module {u'include_role': {u'name': u'role1'}} was not found in configured module paths.
    # - name: Include an entire task
    #   action:
    #     module: "{{ entire_task }}"

    - block:
        - name: Include a role that will fail
          include_role:
            name: role1
            tasks_from: fail.yml

      rescue:
        - name: Include a role inside rescue
          include_role:
            name: role2

      always:
        # Looks like this runs role1, role2, and role3. No idea why.
        - name: Include role inside always
          include_role:
            name: role3
